File: blog.py
# -*- coding: utf-8 -*-

import subprocess
import os
import yaml


# Splits a file in two (three):
#  1. (REMOVED) a first empty file until the first YAML delimiter
#  2. the YAML header. The 1 offset keeps the final delimiter in this block
#  3. the body of the post
# The output files are named with a prefix
file = "testfile.md"
os.system("csplit --keep-files --elide-empty-files "+file+" /---/1 /---/ --prefix=\"post\"")


# Parses the YAML Header
stream  = open("post01","r")
header  = yaml.load(stream)
title   = header["title"].encode('utf-8')
authors = ', '.join(header["authors"]).encode('utf-8')


# Outputs the formatted article
output = """
# {0}
*Autores: {1}*
""".format(title,authors)

outfile = open("post","w")
outfile.write(output)
outfile.close()


# Joins the header with the body
os.system('cat post02 >> post')


# Pandoc renders markdown to latex 
os.system('pandoc post -o post.pdf')

File: blog.rb
#!/usr/bin/env ruby
# Kramdown is the same markdown parser used in our blog
require "kramdown"
# Yaml for reading the front-matter
require "yaml"

# Enable some replacement patterns just in case
md_replacements = {

}
tex_replacements = {
  "\\begin{array}" => "\\[\\begin{array}",
  "\\end{array}"   => "\\end{array}\\]"
}


# Use the argument or assume testfile
file = ARGV[0] || "testfile.md"
# Separate front-matter from post content
`csplit --keep-files --elide-empty-files --suppress-matched #{file} /---/1 /---/ --prefix="post"`

# Read metadata
header = YAML.load File.read("post01")
# Include metadata with rest of content
output = "# #{header["title"]}\n*Autores: #{header["authors"].join(", ")}*\n#{File.read("post02")}"
md_replacements.each{ |search, replace| output.gsub!(search, replace) }
# Convert to LaTeX
latex_output = Kramdown::Document.new(output, template: :document).to_latex
tex_replacements.each{ |search, replace| latex_output.gsub!(search, replace) }
File.write("post.tex", latex_output)
# Convert to PDF and write
`pandoc post.tex -o post.pdf`
